taskmgr/queueservice.go

```diff
@@ -0,0 +1,387 @@
+		// would return an object on the form: {taskId, runId}.
+		m, err := base64.StdEncoding.DecodeString(qm.MessageText)
+		if err != nil {
+			// try to delete from Azure, if it fails, nothing we can do about it
+			// not very serious - another worker will try to delete it
+			q.Log.Errorf("Not able to base64 decode the Message Text '%s' in Azure QueueMessage response.'", qm.MessageText)
+			q.Log.Info("Deleting from Azure queue as other workers will have the same problem.")
+			err := q.deleteFromAzure("", signedDeleteUrl)
+			if err != nil {
+				q.Log.WithFields(logrus.Fields{
+					"url":   signedDeleteUrl,
+					"error": err,
+				}).Warn("Not able to call Azure delete URL")
+			}
+			return nil, err
+		}
+
+		// initialise fields of TaskRun not contained in json string m
+		taskRun := &TaskRun{
+			SignedDeleteUrl: signedDeleteUrl,
+		}
+
+		// now populate remaining json fields of TaskRun from json string m
+		err = json.Unmarshal(m, &taskRun)
+		if err != nil {
+			logger.WithFields(logrus.Fields{
+				"error":   err,
+				"message": m,
+			}).Warn("Not able to unmarshal json from base64 decoded MessageText")
+			err := q.deleteFromAzure("", signedDeleteUrl)
+			if err != nil {
+				logger.WithFields(logrus.Fields{
+					"url":   signedDeleteUrl,
+					"error": err,
+				}).Warn("Not able to call Azure delete URL")
+			}
+			continue
+		}
+		taskRuns = append(taskRuns, taskRun)
+	}
+
+	return taskRuns, nil
+}
+
+// Refreshes a list of task queue urls.  Each task queue contains a pair of signed urls
+// used for polling and deleting messages.
+func (q *queueService) refreshTaskQueueUrls() error {
+	// Attempt to wait until expiration gets closer before refreshing.  No
+	// need to do it more frequently.
+	if !q.shouldRefreshQueueUrls() {
+		return nil
+	}
+
+	q.Log.Debug("Refreshing Azure queue task urls")
+	signedURLs, _, err := q.client.PollTaskUrls(q.ProvisionerId, q.WorkerType)
+	if err != nil {
+		q.Log.WithField("error", err).Warn("Error retrieving task urls.")
+		return errors.New("Error retrieving task urls.")
+	}
+
+	taskQueues := []TaskQueue{}
+	for _, pair := range signedURLs.Queues {
+		taskQueues = append(taskQueues, TaskQueue(pair))
+	}
+
+	q.mu.Lock()
+	q.queues = taskQueues
+	q.Expires = signedURLs.Expires
+	q.mu.Unlock()
+	q.Log.Debugf("Refreshed %d Azure queue task urls", len(taskQueues))
+	return nil
+}
+
+func (q queueService) retrieveTasksFromQueue(ntasks int) (*[]*TaskRun, error) {
+	err := q.refreshTaskQueueUrls()
+	if err != nil {
+		return nil, err
+	}
+
+	tasks := []*TaskRun{}
+	for _, queue := range q.queues {
+		// It hopefully would never be greater, but just incase, we would want to return
+		// and run what tasks we do have.
+		if len(tasks) >= ntasks {
+			return &tasks, nil
+		}
+		taskRuns, err := q.pollTaskUrl(&queue, ntasks-len(tasks), q.Log)
+		if err != nil {
+			q.Log.Warnf("Could not retrieve tasks from the Azure queue. %s", err)
+			continue
+		}
+		tasks = append(tasks, taskRuns...)
```

Comment:
This is not good... It'll work but it's sub-optimal... And increases the risk of timeouts... 
That's why I suggest the alternative interface...

Btw, I propose that the information we get from azure queue is called a `TaskTip` because it's not necessarily a task.. it's not a claim yet. But a _tip_ that there may be a task to claim, so we should attempt to claim it before deleting the tip. But if we get a 409 when trying to claim it we should just delete the obviously invalid tip. 

I see a system as follows: (just pseudo code)
```
//polling thread:
for !self.stopped {
  for self.tasksRequested > 0 && !self.stopped {
    ensureQueueUrls() // Calls queue.PollTaskUrls() if necessary
    for q := range queueUrls {
       // Keep polling from highest level queue until it doesn't return any task
       for !self.stopped {
         taskTips := pollAzureQueue(q, tasksRequested) // returns list of {taskId, runId, signedDeleteUrl}
         if len(taskTips) == 0 {
           break;
         }
         for task := range taskTips {
           go self.claimAndRun();
         }
         self.tasksRequested -= len(taskTips);
       }
       if self.tasksRequested == 0 || self.stopped {
         break;
       }
    }
  }
  // the sleepBetweenPolls is a parameter in the constructor, so the sleep interval can adjust depending on
  // billing cycle (or cloud that we're running on).
  time.Sleep(self.sleepBetweenPolls());
}
```

In the `claimAndRun` something like:
```
// Claim from queue:
resp, err := queue.claimTask(taskId, runId, ...)
if err != nil {
  if err.statusCode == 409 {
    // Call signedDeleteUrl for the taskTip.
  } else {
    log.Error(err);
  }
  self.tasksRequested += 1;
  return;

  // Call signedDeleteUrl for the taskTip as the claimTask was successful
 
  taskClaim := {... resp... taskTip, or combination there of}

  // Again the function runTask is given by the constructor to QueuePoller
  // The worker here supplies a function that takes care of processing the task.
  // and when the worker is done processing the task.. .The worker calls
  // QueuePoller.incTaskRequest(1); Effectively QueuePoller have an internal
  // counter of how many tasks the worker is currently willing to accept.
  // The worker gets these task via. a callback.
  // We could also use a channel, but a callback just seems easier.
  go self.runTask(taskClaim);
}
```

Note: all this psuedo code doesn't account for locking, some places self is missing too... etc...
For some of these things like the `tasksRequested` it might make sense to use a atomic integer, we can easily make one just like I made the AtomicBool (both should probably live in runtime/atomics or something like that).

For `stopped` we should use an atomic bool too...


----
What I'm trying to accomplish here is two things:
 1) When a taskTip is received from azure queue storage, we start a new go routine and start with claimTask, deleteSignedUrl, as well as sending it off to the worker for processing.
 2) If we have two azure queues p1 and p2 and have just polled p1 for 1 task, but didn't get any (we're about to poll p2). But the worker finishes a task and calls `QueuePoller.incTaskRequest(1)` to get a new task for the now empty capacity, p2 will be polled for 2 tasks. And if p1 had returned 1 tasks, but during this request the worker had increased the taskRequest count by one, p1 would be polled again. 
 3) Ensure that the sleeping interval between polls if the polling was empty, is controlled by the worker. (this is also the case in your design as the worker must be doing the sleeping).
 4) If we have two azure queues p1, p2, we request 5 tasks from p1 and get 1, we should poll p1 again, until we get zero tasks, before proceeding to p2. This is a bug in the current implementation. 

(1) is the only one here that is really important. It reduces unnecessary delay and risk of lengthy requests causing azure queue claims to timeout. Note, in the future there will likely be a variable number of priority queues, as datacenter workerTypes might actually build up quite a pending queue. So if there is bad internet weather so retries and polling takes time, the first azure message could risk timing out before we get to call deleteSignedUrl for it. Let's avoid this, and do maximum concurrency when we do I/O, especially when it's time sensitive (even if the sensitivity is on the order of 5 minutes).